servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerTest.java

```diff
@@ -381,6 +400,177 @@ public void newConnectionIsClosedWhenSelectorRejects() throws Exception {
+
+        for (int i = 0; i < DEFAULT_HEALTH_CHECK_FAILED_CONNECTIONS_THRESHOLD; ++i) {
+            Exception e = assertThrows(ExecutionException.class, () -> lb.selectConnection(any()).toFuture().get());
+            assertThat(e.getCause(), instanceOf(DELIBERATE_EXCEPTION.getClass()));
+        }
+
+        for (int i = 0; i < timeAdvancementsTillHealthy; ++i) {
+            unhealthyHostConnectionFactory.advanceTime(testExecutor);
+        }
+
+        final TestLoadBalancedConnection selectedConnection = lb.selectConnection(any()).toFuture().get();
+        assertThat(selectedConnection, equalTo(properConnection.toFuture().get()));
+
+        // 5 failed attempts trigger health check, 2 health check attempts fail, 3rd health check attempt
+        // uses the proper connection, final selection reuses that connection. 8 total creation attempts.
+        int expectedConnectionAttempts = 8;
+        assertThat(unhealthyHostConnectionFactory.requests.get(), equalTo(expectedConnectionAttempts));
+    }
+
+    // Concurrency test, run multiple times (at least 1000).
+    @Test
+    public void hostUnhealthyDoesntRaceToRunHealthCheck() throws Exception {
+        serviceDiscoveryPublisher.onComplete();
+
+        final Single<TestLoadBalancedConnection> properConnection = newRealizedConnectionSingle("address-1");
+        final int timeAdvancementsTillHealthy = 3;
+        final UnhealthyHostConnectionFactory unhealthyHostConnectionFactory = new UnhealthyHostConnectionFactory(
+                "address-1", timeAdvancementsTillHealthy, properConnection);
+        final DelegatingConnectionFactory connectionFactory = unhealthyHostConnectionFactory.createFactory();
+
+        lb = defaultLb(connectionFactory);
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+
+        // Imitate concurrency by running multiple threads attempting to establish connections.
+        ExecutorService executor = Executors.newFixedThreadPool(3);
+        final Runnable runnable = () ->
+                assertThrows(ExecutionException.class, () -> lb.selectConnection(any()).toFuture().get());
+
+        for (int i = 0; i < 1000; i++) {
+            executor.submit(runnable);
+        }
+
+        // From test main thread, wait until the host becomes UNHEALTHY, which is apparent from NoHostAvailableException
+        // being thrown from selection AFTER a health check was scheduled by any thread.
+        try {
+            awaitIndefinitely(lb.selectConnection(any()).retryWhen(retryWithConstantBackoffFullJitter((t) ->
+                    // DeliberateException comes from connection opening, check for that first
+                    // Next, NoAvailableHostException is thrown when the host is unhealthy,
+                    // but we still wait until the health check is scheduled and only then stop retrying.
+                    t instanceof DeliberateException || testExecutor.scheduledTasksPending() == 0,
+            // try to prevent stack overflow
```

Comment:
When I ran this 1000 times, 1 test failed due to stack overflow caused by the nature of resubscribing in the retry operator. I believe adding a small delay will allow the concurrent executor trigger the health check before an overflow happens.